### ISSUE TITLE
Providing viewer interactivity through scripting

### ISSUE THREAD

**Issue by AhmetCanSolak** [napari/napari#520]
## 🚀 Feature

Right now due to nature of qt and our implementation it is not possible to interact with viewer programmatically through scripting. I want to propose providing a way to users to interact with viewer during runtime.

## Motivation

A script would be executed then rendered with the current implementation. But there are many case where one might want to interact with viewer in async fashion.

## Pitch

We can provide a method maybe for users to pass their code and we can call their code. Having a singleton QThreadPool in **napari** could be a nice basis. I have been already using a worker pattern to update viewer in async way. Hence, i am suggesting this. Open to other ways of providing such feature.

## Alternatives

I wont be providing alternatives on this initial comment but i am open to hear alternatives! 

## Additional context

An example script which is not working in the way we want by now:
```python
import numpy as np
from time import sleep
import napari
​
​
def main():
    with napari.gui_qt():
​
        # create an empty viewer
        viewer = napari.Viewer()
​
        # add the array
        layer = viewer.add_image(np.random.random((20, 40)), name='randArray')
​
        for i in range(10):
            layer.data = np.random.random((20, 40))
            print('Updating data...')
            sleep(.5)
​
​
if __name__ == "__main__":
   main()
```

Edit: Current solution for now, thanks to @bryantChhun:
```python
# bchhun, {2019-08-01}
import numpy as np
import napari
from PyQt5.QtCore import QThread
import time
from datetime import datetime
def send_random(layer_):
    # number of images to stream
    num_images = 10
    
    # time between image assignment
    update_interval = 0.5
    # pause so we can observe beginning/end of visual update
    time.sleep(2)
    print('from another thread')
    start = datetime.now()
    for k in range(num_images):
        time.sleep(update_interval)
        dat = np.random.random((512, 512))
        layer_.data = dat
        
        # check that data layer is properly assigned and not blocked?
        while layer.data.all() != dat.all():
            layer_.data = dat
    stop = datetime.now()
    print("time per update = "+
          str((stop-start).microseconds/num_images))
class ThreadRandom(QThread):
    def __init__(self, layer_):
        QThread.__init__(self)
        self.layer_ = layer_
    def run(self):
        send_random(self.layer_)
with napari.gui_qt():
    # create the viewer with an image
    data = np.random.random((512,512))
    viewer = napari.Viewer()
    layer = viewer.add_image(data)
    t = ThreadRandom(layer)
    t.start()
```

**Comment by AhmetCanSolak** [comment]
@sofroniewn i am not happy to see this as a part of 0.3.0 milestone to be honest. I feel there is a significant need. 

Also, I want to suggest providing another contextmanager in napari to provide a way to update viewer. An example:
```python
import numpy as np
from time import sleep
import napari
​
​
def main():
    with napari.gui_qt():
​
        # create an empty viewer
        viewer = napari.Viewer()
​
        # add the array
        layer = viewer.add_image(np.random.random((20, 40)), name='randArray')
​
        # NEW CONTEXTMANAGER
        with napari.update():
               for i in range(10):
                   layer.data = np.random.random((20, 40))
                   print('Updating data...')
                   sleep(.5)
​
​
if __name__ == "__main__":
   main()
```

and we can run everything under update context in a separate thread maybe? What are your opinions on this? @sofroniewn @jni @royerloic

**Comment by kne42** [comment]
Unfortunately a context manager would not work in this scenario since the code is still run. I can not think of any workarounds/magic to make this work without just requiring the user to define a function.

Here is the nicest syntax I can think of:
```python
def main():
    with napari.gui_qt():
        # create an empty viewer
        viewer = napari.Viewer()
​
        # add the array
        layer = viewer.add_image(np.random.random((20, 40)), name='randArray')
​
        # define a new thread to run
        @napari.delayed_thread
        def stream_data():
            for i in range(10):
                layer.data = np.random.random((20, 40))
                print('Updating data...')
                sleep(.5)
```

the thread containing `stream_data` would be scheduled to run on application execution

**Comment by sofroniewn** [comment]
@AhmetCanSolak happy to discuss this more on Thursday and also happy to have a solution for this get merged into `0.2` if we can identify a good one.

I'll change the milestone now, I agree there are a couple people out there including @henrypinkard and Anton Tremsin (can't find his github) who I think will be interested in trying this out right away if we get something working.

**Comment by AhmetCanSolak** [comment]
@kne42 you are right code will still run with context manager. Decorator is a better idea definitely. What do you think about checking if napari running from a jupyter kernel or not in the property setters and handling threading deep in the code? I know will take more work but then people can freely use same syntax on notebooks and scripts.

@sofroniewn thank you so much for quick reply. I am hoping to find a good enough solution to this ASAP. I think we can name more people who needs this feature. Let's see what we can do best.

**Comment by henrypinkard** [comment]
Yes, happy to test whenever it's ready

**Comment by kne42** [comment]
> What do you think about checking if napari running from a jupyter kernel or not in the property setters and handling threading deep in the code?

I was thinking about this too but unfortunately the issue remains in the event that either the data is being live-streamed or any use case where obtaining the data to stream takes a significant amount of time. Even in the same example, it will not work because we are using `sleep` which still blocks the thread.

**Comment by AhmetCanSolak** [comment]
@kne42 what i mean was in the setter doing below:
```python
if running_as_script:
    fn = normal_setter_function
    t = Thread(fn)
    t.start()
``` 

so user will not pass any function with some sleep commands to setter. We could adjust the property setter in a way to execute from a separate thread, if user wants sleep calls in their script that will be fine.

**Comment by kne42** [comment]
@AhmetCanSolak could you provide a hypothetical example script where this is implemented as described?

**Comment by AhmetCanSolak** [comment]
@kne42 as we discussed offline my example script not achieving what we want. Seems letting users to write their methods still seems as best option

**Comment by sofroniewn** [comment]
@AhmetCanSolak can this be closed now that #650 is merged?

**Comment by AhmetCanSolak** [comment]
yes @sofroniewn thank you for heads up!
